src/frontend/lib/storage/provider.ts

```diff
@@ -0,0 +1,204 @@
+                method: 'POST',
+                headers: { 'Content-Type': 'application/json' },
+                body: JSON.stringify(requestBody),
+            })
+
+            if (!response.ok) {
+                const errorData = await response.json()
+                console.error('Presigned URL request failed:', {
+                    status: response.status,
+                    error: errorData,
+                })
+                throw new Error(errorData.details || 'Failed to get upload URL')
+            }
+
+            const data = await response.json()
+            console.log('Received presigned URL data:', data, {
+                key: data.key,
+                expiresIn: data.expiresIn,
+            })
+
+            return data
+        } catch (error) {
+            console.error('Error getting presigned URL:', error)
+            throw error
+        }
+    }
+
+    private async uploadWithProgress(
+        url: string,
+        file: File,
+        onProgress?: (progress: UploadProgress) => void,
+    ): Promise<Response> {
+        return new Promise((resolve, reject) => {
+            const xhr = new XMLHttpRequest()
+
+            xhr.upload.addEventListener('progress', event => {
+                if (event.lengthComputable && onProgress) {
+                    onProgress({
+                        loaded: event.loaded,
+                        total: event.total,
+                        percentage: (event.loaded / event.total) * 100,
+                    })
+                }
+            })
+
+            xhr.addEventListener('load', () => {
+                const isValidStatus = xhr.status >= 200 && xhr.status < 300
+
+                if (isValidStatus) {
+                    resolve(
+                        new Response(xhr.response, {
+                            status: xhr.status,
+                            headers:
+                                this.config.provider !== Provider.Azure
+                                    ? new Headers({
+                                          ETag:
+                                              xhr.getResponseHeader('ETag') ||
+                                              '',
+                                      })
+                                    : undefined,
+                        }),
+                    )
+                } else {
+                    reject(
+                        new Error(
+                            `Upload failed with status ${xhr.status}: ${xhr.responseText}`,
+                        ),
+                    )
+                }
+            })
+
+            xhr.addEventListener('error', () =>
+                reject(new Error('Network error during upload')),
+            )
+
+            xhr.open('PUT', url)
+            if (this.config.provider === Provider.Azure)
+                xhr.setRequestHeader('x-ms-blob-type', 'BlockBlob')
+            xhr.send(file)
+        })
+    }
+
+    validateConfig(): boolean {
```

Comment:
Validation of Required Configurations:

Issue: The validateConfig method only checks for tokenEndpoint, but provider is also critical for the functionality.

Recommendation: Update the validateConfig method to include provider:

validateConfig(): boolean {
    const required = ['tokenEndpoint', 'provider'] as const;
    const missing = required.filter(key => !this.config[key]);

    if (missing.length > 0) {
        throw new Error(`Missing required configuration: ${missing.join(', ')}`);
    }

    // Optionally, validate that provider is a valid enum value
    if (!Object.values(Provider).includes(this.config.provider)) {
        throw new Error(`Invalid provider: ${this.config.provider}`);
    }

    return true;
}

Reply:
Thanks @BSalaeddin  done